lib/gears/reactive.lua

```diff
@@ -0,0 +1,647 @@
+    assert(
+        type(up) == "table",
+        "Trying to index a "..type(up).." value in a `gears.reactive` expression."
+    )
+
+    local upk = up[key]
+
+    -- Connect.
+    add_key(self, key)
+
+    local ret, is_proxy = create_proxy(upk, self._reactive.origin)
+
+    -- Always query the non-proxy upvalue. We cannot detect if they
+    -- change.
+    if is_proxy then
+        rawset(self, key, ret)
+    end
+
+    return ret
+end
+
+-- Set valuers trough the proxy.
+local function proxy_newindex(self, key, value)
+    rawset(self, key, create_proxy(value, self._reactive.origin))
+
+    -- Strip the proxy before setting the value on the original object.
+    if type(value) == "table" and value._reactive then
+        value = get_upvalue(value)
+    end
+
+    local v = get_upvalue(self)
+
+    v[key] = value
+
+    -- Connect.
+    if is_gobject(v) and not self._reactive.origin.objects[self][key] then
+        add_key(self, key)
+    end
+end
+
+-- It's possible that multiple proxy points to the same value.
+-- While we could make a large map of all proxies, it's simpler
+-- to just implement __eq/__lt/__le and be done.
+local function proxy_equal(a, b)
+    a, b = get_upvalue(a), get_upvalue(b)
+
+    return a == b
+end
+
+local function proxy_lesser(a, b)
+    a, b = get_upvalue(a), get_upvalue(b)
+
+    return a < b
+end
+
+local function proxy_lesser_equal(a, b)
+    a, b = get_upvalue(a), get_upvalue(b)
+
+    return a <= b
+end
+
+local function proxy_tostring(o)
+    return tostring(get_upvalue(o))
+end
+
+-- Wrap tables and functions into a proxy object.
+create_proxy = function(input, origin, getter)
+    local t = type(input)
+    local is_call = is_callable(input)
+
+    -- Everything but the tables are immutable.
+    if t ~= "table" and not is_call then return input, false end
+
+    -- Remove any foreign proxy.
+    if t ~= "function" and input._reactive and input._reactive.origin ~= origin then
+        input = get_upvalue(input)
+    end
+
+    return setmetatable({_reactive = gen_proxy_md(input, origin, getter)}, {
+        __index    = proxy_index,
+        __newindex = proxy_newindex,
+        __call     = proxy_call,
+        __eq       = proxy_equal,
+        __lt       = proxy_lesser,
+        __le       = proxy_lesser_equal,
+        __tostring = proxy_tostring
+    }), true
+end
+
+-- Create a "fake" upvalue ref because otherwise the sandbox could
+-- "leak" and set upvalue in actual code outside of the function.
+-- By "leak", I mean the proxy would become visible from outside
+-- of the sandbox by the reactive expression "siblings" sharing the
+-- same environment.
+local function copy_upvalue_reference(loaded, idx, value)
+    -- Make sure it has a fresh upvalue reference. Something that
+    -- we are sure cannot be shared with something else.
+    local function fake_upvalue_env()
+        return value
+    end
+
+    debug.upvaluejoin(loaded, idx, fake_upvalue_env, 1) --luacheck: ignore
```

Comment:
I definitely do not understand all the magic here, but... by breaking upvalues, could you not break code / cause hard to debug issues? The "function in the sandbox" changed some upvalue but that change is not visible to outside code.

Basically... all writes to upvalues must be forbidden...?

Reply:
I think you get it as per the comment below, but it's "fine" to write the upvalue. `string.dump`/`load` make a copy of the function, then this code makes a copy of the upvalue "vtable". So the lambda function above has a single upvalue (since it has almost no code and the Lua doesn't put all possible upvalues in the vtable. Then I take the *reference* to the lambda upvalue and put it in the target function. This means that when `loaded` accesses that upvalue, it actually "live" in the lambda environment. If you set something, then the value of... `value` in the lambda would change (assuming it has not been GC-ed already), which is harmless. From the point of view of `loaded`, the value was really set. From the point of view of the "real world", nothing happened.